guides/building-a-middleware-from-scratch.md

```diff
@@ -0,0 +1,693 @@
+# Building a middleware from scratch
+
+In ["Inventing the `Service` trait"][invent] we learned all the motivations
+behind [`Service`] and why its designed the way it is. We also built a few
+smaller middleware ourselves but we took a few shortcuts in our implementation.
+In this guide we're going to build the `Timeout` middleware as it exists in
+Tower today without taking any shortcuts.
+
+Writing a robust middleware requires working with async Rust at a slightly lower
+level than you might be used to. The goal of this guide is to demystify the
+concepts and patterns so you can start writing your own middleware and maybe
+even contribute back to the Tower ecosystem!
+
+## Getting started
+
+The middleware we're going to build is [`tower::timeout::Timeout`]. It will set
+a limit on the maximum duration its inner `Service` is allowed to take. If it
+doesn't produce a response within some amount of time, an error is returned.
+This allows the client to retry that request or report an error to the user,
+rather than waiting forever.
+
+Lets start by writing a `Timeout` struct that holds the `Service` its wrapping
+and the duration of the timeout:
+
+```rust
+use std::time::Duration;
+
+struct Timeout<T> {
+    inner: T,
+    timeout: Duration,
+}
+```
+
+As we learned in ["Inventing the `Service` trait"][invent] its important for
+services to implement `Clone` such that you can convert the `&mut self` given to
+`Service::call` into an owned `self` that can be moved into the response future,
+if necessary. We should therefore add `#[derive(Clone)]` to our struct. We
+should also drive `Debug` while we're at it:
+
+```rust
+#[derive(Debug, Clone)]
```

Comment:
I think this is common enough in the ecosystem that I probably wouldn't even include text on why we're choosing to do it. See also [the Rust API guidelines' recommendation](https://rust-lang.github.io/api-guidelines/interoperability.html#types-eagerly-implement-common-traits-c-common-traits).

Reply:
Hm yeah thats true. I think I'm going to leave it though since its extra important for services so users can move the service into the response future.